go/nbs/conjoiner.go

```diff
@@ -0,0 +1,201 @@
+		}
+		c.waiters[id] = make(chan struct{})
+		go c.runAndNotify(id, f)
+		return c.waiters[id]
+	}()
+	if testWg != nil {
+		testWg.Done()
+	}
+	<-wait
+}
+
+// runAndNotify runs f() and, upon completion, signals everyone who called
+// await(id).
+func (c *asyncConjoiner) runAndNotify(id string, f func()) {
+	f()
+
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	ch, present := c.waiters[id]
+	d.PanicIfFalse(present)
+	close(ch)
+	delete(c.waiters, id)
+}
+
+func conjoin(mm manifest, p tablePersister, needsConjoin func(int) bool, stats *Stats) {
+	b := &backoff.Backoff{
+		Min:    128 * time.Microsecond,
+		Max:    10 * time.Second,
+		Factor: 2,
+		Jitter: true,
+	}
+
+	exists, _, lock, root, upstream := mm.ParseIfExists(nil)
+	d.PanicIfFalse(exists)
+	// This conjoin may have been requested by someone with an out-of-date notion of what's upstream. Verify that we actually still believe a conjoin is needed and, if not, return early
+	if !needsConjoin(len(upstream)) {
+		return
+	}
+
+	var conjoined tableSpec
+	var conjoinees, keepers []tableSpec
+
+	for {
+		if conjoinees == nil {
+			conjoined, conjoinees, keepers = conjoinTables(p, upstream, stats)
+		}
+
+		specs := append(make([]tableSpec, 0, len(keepers)+1), conjoined)
+		specs = append(specs, keepers...)
+
+		nl := generateLockHash(root, specs)
+		lock, root, upstream = mm.Update(lock, nl, specs, root, nil)
+
+		if nl == lock {
+			return
+		}
+		// Optimistic lock failure. Someone else moved to the root, the set of tables, or both out from under us.
+		// If we can re-use the conjoin we already performed, we want to try again. Currently, we will only do so if ALL conjoinees are still present upstream. If we can't re-use...then someone else almost certainly landed a conjoin upstream. In this case, bail and let clients ask again if they think they still can't proceed.
+		conjoineeSet := map[addr]struct{}{}
+		upstreamNames := map[addr]struct{}{}
+		for _, spec := range upstream {
+			upstreamNames[spec.name] = struct{}{}
+		}
+		for _, c := range conjoinees {
+			if _, present := upstreamNames[c.name]; !present {
+				return // Bail!
+			}
+			conjoineeSet[c.name] = struct{}{}
+		}
+
+		// Filter conjoinees out of upstream to generate new set of keepers
+		keepers = make([]tableSpec, 0, len(upstream)-len(conjoinees))
+		for _, spec := range upstream {
+			if _, present := conjoineeSet[spec.name]; !present {
+				keepers = append(keepers, spec)
+			}
+		}
+		time.Sleep(b.Duration())
+	}
+}
+
+func conjoinTables(p tablePersister, upstream []tableSpec, stats *Stats) (conjoined tableSpec, conjoinees, keepers []tableSpec) {
+	// Open all the upstream tables concurrently
+	sources := make(chunkSources, len(upstream))
+	wg := sync.WaitGroup{}
+	for i, spec := range upstream {
+		wg.Add(1)
+		go func(idx int, spec tableSpec) {
+			sources[idx] = p.Open(spec.name, spec.chunkCount)
+			wg.Done()
+		}(i, spec)
+		i++
+	}
+	wg.Wait()
+
+	t1 := time.Now()
+
+	sortedUpstream := make(chunkSources, len(sources))
```

Comment:
maybe factor out the code which decides *which* tables to pick for conjoinment into another function which has a comment describing the approach (i.e. choose the smallest N tables which when removed and replaced with the conjoinment, will still be the smallest table).